kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiDefaultStream.java

```diff
@@ -0,0 +1,158 @@
+                    attempt++;
+                }
+            } while (attempt < (retry + additionalRetriesForTableCreation));
+            throw new BigQueryStorageWriteApiConnectException(
+                    String.format(
+                            "Exceeded %s attempts to create Default stream writer on table %s ",
+                            (retry + additionalRetriesForTableCreation), tableName
+                    ), mostRecentException);
+        });
+    }
+
+    /**
+     * Calls AppendRows and handles exception if the ingestion fails
+     * @param tableName  The table to write data to
+     * @param rows       The records to write
+     * @param streamName The stream to use to write table to table. This will be DEFAULT always.
+     */
+    @Override
+    public void appendRows(TableName tableName, List<Object[]> rows, String streamName) {
+        JSONArray jsonArr = new JSONArray();
+        JsonStreamWriter writer = getDefaultStream(tableName.toString());
+
+        rows.forEach(item -> jsonArr.put(item[1]));
+
+        logger.debug("Sending {} records to write Api default stream on {} ...", rows.size(), tableName);
+        int attempt = 0;
+        Exception mostRecentException = null;
+        boolean tableCreationAttempted = false;
+        int additionalRetriesForTableCreation = 0;
+        do {
+            try {
+                if (attempt > 0) {
+                    waitRandomTime();
+                }
+                logger.trace("Sending records to Storage API writer...");
+                ApiFuture<AppendRowsResponse> response = writer.append(jsonArr);
+                AppendRowsResponse writeResult = response.get();
+                logger.trace("Received response from Storage API writer...");
+
+                if (writeResult.hasUpdatedSchema()) {
+                    logger.warn("Sent records schema does not match with table schema, will attempt to update schema");
+                    //TODO: Update schema attempt Once
+                } else if (writeResult.hasError()) {
+                    Status errorStatus = writeResult.getError();
+                    String errorMessage = String.format("Failed to write rows on table %s due to %s", tableName, errorStatus.getMessage());
+                    if (BigQueryStorageWriteApiErrorResponses.isMalformedRequest(errorStatus.getCode())) {
+                        // Fail on data error
+                        throw new BigQueryStorageWriteApiConnectException(tableName.getTable(), writeResult.getRowErrorsList());
+                        //TODO: DLQ handling
+                    } else if (!BigQueryStorageWriteApiErrorResponses.isRetriableError(errorStatus.getMessage())) {
+                        // Fail on non-retriable error
+                        logger.error(errorMessage);
+                        throw new BigQueryStorageWriteApiConnectException(errorMessage);
+                    }
+                    logger.warn(errorMessage + "Retrying...");
```

Comment:
Lets also log which attempt is being retried here.

Reply:
Thanks for the suggestion, will add.